### ISSUE TITLE
Query on segmentation training script

### ISSUE THREAD

**Issue by innat** [keras-team/keras-cv#987]
- why we need to do [this](https://github.com/keras-team/keras-cv/blob/master/examples/training/semantic_segmentation/pascal_voc/deeplab_v3.py#L88-L90)?
- If the segmentation map is integer encoded mask, will be there any class or (anomaly pixel value) represented by 255 as we [ingored](https://github.com/keras-team/keras-cv/blob/master/examples/training/semantic_segmentation/pascal_voc/deeplab_v3.py#L129) it in metrices?
- The [DeepLabV3 ](https://github.com/keras-team/keras-cv/blob/a402f614c48924ebcfddbec6a57a18dbfc21f1ab/keras_cv/models/segmentation/deeplab.py#L31) segmentation model (and also obj. detection) is written in subclass API (unlike classification model). By this `summary` and `plot_model` are no longer intuitive to get, same as classification models. 
- How many epochs is used to get ~63 mIOU? Is it 100 (last epoch)? I did some change in data loader and without augmentation, I able to achieve over 64+ mIOU on validation score roughly around 10 epoch and also the val_miou score was decently increasing per epoch.

---

findings
- multi-gpu training makes val scores to `nan`, tested on Tesla T4 gcp, TF 2.10.
- `jit_compile = True` in compiler gives error.
- tpu test: todo

cc @tanzhenyu

**Comment by tanzhenyu** [comment]
255 is being used to ignore boundary classes.
not sure why T4 would give nan, I mostly use V100 multi-gpu training
`jit_compile` / tpu has not been tested yet, I also need to test mixed precision.

**Comment by innat** [comment]
Thanks for the response. Could you please elaborate on boundary classes issue. I didn't make an eda on SBD dataset. Is the boundary classes (255, white pixel) present in the segmentaiton mask in this dataset? While using the training script, I opted out this operation. 

I modified the data loader a bit as follows. And also I didn't set any `ignore_class` in loss and metrics.

```python
def load_image(datapoint):
    input_image = tf.image.resize(
        datapoint['image'], (image_size, image_size)
    )
    input_image = tf.cast(input_image, dtype=tf.float32)

    input_mask = tf.image.resize(
        datapoint['class_segmentation'], (image_size, image_size),
        method=tf.image.ResizeMethod.NEAREST_NEIGHBOR
      )
    input_mask = tf.cast(input_mask, tf.uint8)
    return input_image, input_mask
```
```python
train_ds = load(split="sbd_train", data_dir=None)
eval_ds = load(split="sbd_eval", data_dir=None)

train_datagen = train_ds.map(load_image, num_parallel_calls=tf.data.AUTOTUNE)
train_datagen = train_datagen.batch(global_batch)
train_datagen = train_datagen.shuffle(8 * global_batch)
train_datagen = train_datagen.prefetch(buffer_size=tf.data.AUTOTUNE)

eval_datagen = eval_ds.map(load_image, num_parallel_calls=tf.data.AUTOTUNE)
eval_datagen = eval_datagen.batch(global_batch, drop_remainder=True)
eval_datagen = eval_datagen.prefetch(buffer_size=tf.data.AUTOTUNE)
```

---

I aslo tested with mixed precision, works fine. About augmentation, I tried to adopt some functionality from [here](https://www.tensorflow.org/tutorials/images/segmentation) `Augment` class, but for some reason, the `labels` wasn't augmenting same as its pair. Not sure if its somehow bug in TF 2.10, (will re-check).

```python
def call(self, inputs, labels):
    inputs = self.augment_inputs(inputs)
    labels = self.augment_labels(labels)
    return inputs, labels
```

**Comment by tanzhenyu** [comment]
> Thanks for the response. Could you please elaborate on boundary classes issue. I didn't make an eda on SBD dataset. Is the boundary classes (255, white pixel) present in the segmentaiton mask in this dataset? While using the training script, I opted out this operation.
> 
> I modified the data loader a bit as follows. And also I didn't set any `ignore_class` in loss and metrics.
> 
> ```python
> def load_image(datapoint):
>     input_image = tf.image.resize(
>         datapoint['image'], (image_size, image_size)
>     )
>     input_image = tf.cast(input_image, dtype=tf.float32)
> 
>     input_mask = tf.image.resize(
>         datapoint['class_segmentation'], (image_size, image_size),
>         method=tf.image.ResizeMethod.NEAREST_NEIGHBOR
>       )
>     input_mask = tf.cast(input_mask, tf.uint8)
>     return input_image, input_mask
> ```
> 
> ```python
> train_ds = load(split="sbd_train", data_dir=None)
> eval_ds = load(split="sbd_eval", data_dir=None)
> 
> train_datagen = train_ds.map(load_image, num_parallel_calls=tf.data.AUTOTUNE)
> train_datagen = train_datagen.batch(global_batch)
> train_datagen = train_datagen.shuffle(8 * global_batch)
> train_datagen = train_datagen.prefetch(buffer_size=tf.data.AUTOTUNE)
> 
> eval_datagen = eval_ds.map(load_image, num_parallel_calls=tf.data.AUTOTUNE)
> eval_datagen = eval_datagen.batch(global_batch, drop_remainder=True)
> eval_datagen = eval_datagen.prefetch(buffer_size=tf.data.AUTOTUNE)
> ```
> 
> I aslo tested with mixed precision, works fine. About augmentation, I tried to adopt some functionality from [here](https://www.tensorflow.org/tutorials/images/segmentation) `Augment` class, but for some reason, the `labels` wasn't augmenting same as its pair. Not sure if its somehow bug in TF 2.10, (will re-check).
> 
> ```python
> def call(self, inputs, labels):
>     inputs = self.augment_inputs(inputs)
>     labels = self.augment_labels(labels)
>     return inputs, labels
> ```

So basically 255 are coming from labels (segmentation map), they represent something outside of [0, num_foreground_classes] which you don't need to train, so we could mask them out, or cast them to 0 (as background class)

**Comment by innat** [comment]
( Sorry for these many queries, you may be still working to update the script. I'm currently working on a project where segmentation is involved. And I want to use kcv in the pipeline if possible. )

---

> So basically 255 are coming from labels (segmentation map), they represent something outside of [0, num_foreground_classes] 

Yes, understandable. It's an unwanted situation in segmentation mask. I think it's not general for all seg dataset but appear sometimes, for example, in [Stanford Background](https://www.kaggle.com/datasets/ipythonx/stanford-background-dataset) dataset, some mask appears as follows (255 pixel value)

![__results___7_6](https://user-images.githubusercontent.com/17668390/200999973-011585a3-b0a2-430d-9321-28507cb6d9fc.png)

---

This [lines](https://github.com/keras-team/keras-cv/blob/master/examples/training/semantic_segmentation/pascal_voc/deeplab_v3.py#L88-L90) can be replaced with the following, :)

```python
input_mask = tf.image.resize(
        datapoint['class_segmentation'], (image_size, image_size),
        method=tf.image.ResizeMethod.NEAREST_NEIGHBOR
      )
input_mask = tf.where(input_mask == 255, np.dtype('uint8').type(0), input_mask)
input_mask = tf.cast(input_mask, tf.uint8)
```

**Comment by tanzhenyu** [comment]
> ( Sorry for these many queries, you may be still working to update the script. I'm currently working on a project where segmentation is involved. And I want to use kcv in the pipeline if possible. )
> 
> > So basically 255 are coming from labels (segmentation map), they represent something outside of [0, num_foreground_classes]
> 
> Yes, understandable. It's an unwanted situation in segmentation mask. I think it's not general for all seg dataset but appear sometimes, for example, in [Stanford Background](https://www.kaggle.com/datasets/ipythonx/stanford-background-dataset) dataset, some mask appears as follows (255 pixel value)
> 
> ![__results___7_6](https://user-images.githubusercontent.com/17668390/200999973-011585a3-b0a2-430d-9321-28507cb6d9fc.png)
> 
> This [lines](https://github.com/keras-team/keras-cv/blob/master/examples/training/semantic_segmentation/pascal_voc/deeplab_v3.py#L88-L90) can be replaced with the following, :)
> 
> ```python
> input_mask = tf.image.resize(
>         datapoint['class_segmentation'], (image_size, image_size),
>         method=tf.image.ResizeMethod.NEAREST_NEIGHBOR
>       )
> input_mask = tf.where(input_mask == 255, np.dtype('uint8').type(0), input_mask)
> input_mask = tf.cast(input_mask, tf.uint8)
> ```

I see. So would you want `ignore_mask` to be configurable?
